api/src/main/java/com/epam/pipeline/manager/datastorage/providers/aws/s3/S3Helper.java

```diff
@@ -300,6 +303,53 @@ public void restoreFileVersion(String bucket, String path, String version) {
         moveS3Object(client, bucket, new MoveObjectRequest(path, version, path));
     }
 
+    public void restoreFolder(String bucket, String path, RestoreFolderVO restoreFolderVO) {
+        AmazonS3 client = getDefaultS3Client();
+        String requestPath = Optional.ofNullable(path).orElse("");
+        if (!StringUtils.isNullOrEmpty(requestPath)) {
+            Assert.isTrue(checkItemType(client, bucket, requestPath, true) == DataStorageItemType.Folder,
+                    messageHelper.getMessage(MessageConstants.ERROR_FOLDER_INVALID_PATH, requestPath));
+        }
+        try (S3ObjectDeleter deleter = new S3ObjectDeleter(client, bucket)) {
+            restoreFolderFiles(client, bucket, requestPath, restoreFolderVO, deleter);
+        } catch (SdkClientException e) {
+            throw new DataStorageException(e.getMessage(), e.getCause());
+        }
+    }
+
+    private void restoreFolderFiles(AmazonS3 client, String bucket, String path, RestoreFolderVO restoreFolderVO,
+                                    S3ObjectDeleter deleter) {
+        getDeletedFiles(client, bucket, ProviderUtils.withTrailingDelimiter(path), restoreFolderVO, deleter)
+                .forEach(item -> deleter.deleteKey(item.getPath(), ((DataStorageFile) item).getVersion()));
+    }
+
+    private List<AbstractDataStorageItem> getDeletedFiles(AmazonS3 client, String bucket, String path,
+                                                          RestoreFolderVO restoreFolderVO, S3ObjectDeleter deleter) {
+        return listVersions(client, bucket, path, null, null).getResults().stream()
+                .peek(item -> recursiveRestoreFolderCall(item, client, bucket, restoreFolderVO, deleter))
+                .filter(item -> isFileDeletedAndShouldBeRestore(item, restoreFolderVO))
+                .collect(Collectors.toList());
+    }
+
+    private void recursiveRestoreFolderCall(AbstractDataStorageItem item, AmazonS3 client, String bucket,
+                                            RestoreFolderVO restoreFolderVO, S3ObjectDeleter deleter) {
+        if (item.getType() == DataStorageItemType.Folder && restoreFolderVO.isRecursively()) {
+            restoreFolderFiles(client, bucket, item.getPath(), restoreFolderVO, deleter);
+        }
+    }
+
+    private Boolean isFileDeletedAndShouldBeRestore(AbstractDataStorageItem item, RestoreFolderVO restoreFolderVO) {
+        final AntPathMatcher matcher = new AntPathMatcher();
+        return item.getType() == DataStorageItemType.File &&
+                ((DataStorageFile) item).getDeleteMarker() &&
+                isFileFromRestoreList(restoreFolderVO.getIncludeList(), pattern -> matcher.match(pattern, item.getName())) &&
+                isFileFromRestoreList(restoreFolderVO.getExcludeList(), pattern -> !matcher.match(pattern, item.getName()));
+    }
+
+    private Boolean isFileFromRestoreList(List<String> includeOrExcludeList, Predicate<String> pattern) {
```

Comment:
Could the return value be the primitive?

Reply:
fixed